File: src/examples/basic.cpp
#include <iostream>
#include "../ganglion.h"

void handle_message(void * consumer, char * payload, int length, int partition, long offset) {
  printf("Received message with offset %ld from partition %d\n", offset, partition);
}

void handle_report(void * producer, const char * topic, char * payload, int length) {
  printf("Successfully sent message to topic: %s\n", topic);
}

int main(void) {
  GanglionSupervisor *supervisor = new GanglionSupervisor();
  GanglionConsumer *consumer_one = new GanglionConsumer("localhost:9092", 400, "analytics", "libganglion.cpp.analytics", handle_message);
  GanglionConsumer *consumer_two = new GanglionConsumer("localhost:9092", 400, "stream", "libganglion.cpp.analytics", handle_message);

  std::cout << "Adding Consumers" << std::endl;

  supervisor->add(consumer_one);
  supervisor->add(consumer_two);

  std::cout << "Starting Supervisor" << std::endl;

  supervisor->start();

  GanglionProducer *producer = new GanglionProducer("localhost:9092", "example", "none", 1000, 100, handle_report);

  std::string str = "hello";
  std::vector<char> data(str.begin(), str.end());
  producer->publish("analytics", data);
  producer->publish("stream", data);

  std::cin.ignore();

  if (supervisor->is_started()) {
    supervisor->stop();
  }

  delete consumer_one;
  delete consumer_two;
  delete supervisor;

  return 0;
}

File: rust/examples/basic.rs
extern crate ganglion;

use ganglion::{GanglionSupervisor, GanglionProducer, GanglionConsumer, GanglionConsumerHandler};
use std::thread;
use std::time::Duration;

use std::sync::{Arc, Mutex};

#[derive(Clone)]
struct ConsumerHandler {
    count: Arc<Mutex<i32>>
}

impl ConsumerHandler {
    pub fn new() -> ConsumerHandler {
        ConsumerHandler { count: Arc::new(Mutex::new(0)) }
    }
}

impl GanglionConsumerHandler for ConsumerHandler {
    fn handle_message(&mut self, _: i32, _: i64, topic: String, _: Vec<u8>) {
        let mut count = self.count.lock().unwrap();
        *count += 1;
        println!("Received {} message(s) for topic: '{}'", *count, topic);
    }
}

fn main() {
    let mut producer = GanglionProducer::new("localhost:9092", "rust_ganglion", "none", 1000, 100);

    let mut supervisor = GanglionSupervisor::new();
    println!("Making consumers");

    let consumer_one = GanglionConsumer::new("localhost:9092", 100, "analytics", "rust_ganglion.analytics", ConsumerHandler::new());
    let consumer_two = GanglionConsumer::new("localhost:9092", 100, "stream", "rust_ganglion.analytics", ConsumerHandler::new());

    println!("Registering consumers");

    supervisor.register(consumer_one);
    supervisor.register(consumer_two);

    println!("Starting consumers");

    supervisor.start();

    thread::sleep(Duration::from_millis(3000));

    println!("Sending messages");

    producer.publish("analytics", b"foo!!!");
    producer.publish("analytics", b"bar!!!");
    producer.publish("stream", b"foo!!!");
    producer.publish("stream", b"bar!!!");

    thread::sleep(Duration::from_millis(1000));

    supervisor.stop();
}
